src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs

```diff
@@ -77,22 +81,73 @@ private void AcceptFailed(SocketError error, string? exceptionMessage)
         }
 
         [NonEvent]
-        public void ConnectStart(SocketAddress address)
+        public Activity? ConnectStart(SocketAddress address, EndPoint endPoint,  bool keepActivityCurrent)
         {
             Interlocked.Increment(ref _currentOutgoingConnectAttempts);
 
             if (IsEnabled(EventLevel.Informational, EventKeywords.All))
             {
                 ConnectStart(address.ToString());
             }
+
+            Activity? activity = null;
+            if (s_connectActivitySource.HasListeners())
+            {
+                Activity? activityToReset = keepActivityCurrent ? Activity.Current : null;
+                activity = s_connectActivitySource.StartActivity(ConnectActivityName, ActivityKind.Client);
+                if (keepActivityCurrent)
+                {
+                    // Do not overwrite Activity.Current in the caller's ExecutionContext.
+                    Activity.Current = activityToReset;
+                }
+            }
+
+            if (activity is not null)
+            {
+                if (endPoint is IPEndPoint ipEndPoint)
+                {
+                    string peerAddress = ipEndPoint.Address.ToString();
+                    int port = ipEndPoint.Port;
+                    activity.DisplayName = $"socket connect {peerAddress}:{port}";
```

Comment:
Nit: if `IsAllDataRequested` is expected to be rare, you could save a string allocation by using `{ipEndPoint.Address}` directly here (assuming we aren't already allocating the `ToString` elsewhere).